src/widgets/dialogs/LogsPopup.cpp

```diff
@@ -0,0 +1,142 @@
+LogsPopup::LogsPopup()
+{
+    this->initLayout();
+    this->resize(400, 600);
+}
+
+void LogsPopup::initLayout()
+{
+    {
+        QVBoxLayout *layout = new QVBoxLayout(this);
+        layout->setMargin(0);
+
+        {
+            this->channelView = new ChannelView(this);
+            layout->addWidget(this->channelView);
+        }
+
+        this->setLayout(layout);
+    }
+}
+
+void LogsPopup::setInfo(ChannelPtr channel, QString username)
+{
+    this->channel_ = channel;
+    this->userName = username;
+    this->setWindowTitle("Logs for " + userName + " in " + channel->name + "'s history");
+    this->getLogviewerLogs();
+}
+
+void LogsPopup::setupView()
+{
+    ChannelPtr logsChannel(new Channel("logs", Channel::None));
+    QStringList output = this->answer.split("<br>", QString::SkipEmptyParts);
+    std::vector<MessagePtr> messages;
+
+    for (size_t i = 0; i < output.size(); i++) {
+        if (this->usedLogviewer) {
+            MessageParseArgs args;
+            auto ircMessage = Communi::IrcMessage::fromData(output[i].toUtf8(), nullptr);
+            auto privMsg = static_cast<Communi::IrcPrivateMessage *>(ircMessage);
+            TwitchMessageBuilder builder(logsChannel.get(), privMsg, args);
+            if (!builder.isIgnored()) {
+                messages.push_back(builder.build());
+            }
+        } else {
+            MessagePtr message(new Message);
+            message->addElement(new TimestampElement(this->rustleTime));
+            message->addElement(
+                new TextElement(this->userName, MessageElement::Username, MessageColor::System));
+            message->addElement(
+                new TextElement(output[i], MessageElement::Text, MessageColor::Text));
+            messages.push_back(message);
+        }
+    }
+    logsChannel->addMessagesAtStart(messages);
+    this->channelView->setChannel(logsChannel);
+}
+
+void LogsPopup::getLogviewerLogs()
+{
+    TwitchChannel *twitchChannel = dynamic_cast<TwitchChannel *>(this->channel_.get());
+
+    QString channelName = twitchChannel->name;
+    this->channelName_ = channelName;
+
+    QString url = QString("https://cbenni.com/api/logs/%1/?nick=%2&before=500")
+                      .arg(channelName, this->userName);
+
+    NetworkRequest req(url);
+    req.setCaller(QThread::currentThread());
+
+    req.onError([this](int errorCode) {
+        this->getOverrustleLogs();
+        return true;
+    });
+
+    req.getJSON([this, channelName](QJsonObject &dataCbenni) {
+        QJsonValue before = dataCbenni.value("before");
+        for (int i = before.toArray().size() - 1; i >= 0; --i) {
+            int index = before.toArray().size() - 1 - i;
+            QString message = before[index]["text"].toString();
+            // Hacky way to fix the timestamp
+            message.insert(1, "historical=1;");
+            message.insert(1, QString("tmi-sent-ts=%10000;").arg(before[index]["time"].toInt()));
+            this->answer += message + "<br>";
```

Comment:
Create a `std::vector<MessagePtr>` here instead of storing the answer as a member.